### PR TITLE
abort_source: request_abort: restore legacy no-args method

### PR DESCRIPTION
**Pull request by bhalevy**
Change 19b9e4d2c9b5cc6834b5df928092cd7d820b0700 added an optional
parameter to request_abort, defaulted to nullopt.

Normally, this is backward compatible, but sites using
the no-args method by reference / pointer break since
it is no longer available.
For example: `_abort_sources.invoke_on_all(&abort_source::request_abort)`

This change restores the legacy method with the no-args signature
and renames the additional entry-points that accept an exception
to request_abort_ex to keep the interface backward compatible.
Otherwise, the compiler is not smart enough to derive
the right overload for invoke_on_all. See following error from
scylla/main.c:
```
main.cc:117:35: error: no matching member function for call to 'invoke_on_all'
            return _abort_sources.invoke_on_all(&abort_source::request_abort);
                   ~~~~~~~~~~~~~~~^~~~~~~~~~~~~
/home/bhalevy/dev/scylla/seastar/include/seastar/core/sharded.hh:214:14: note: candidate function not viable: no overload of 'request_abort' matching 'std::function<future<> (abort_source &)>' for 1st argument
```

Adjusted unit test to use the renamed method and to exercise
them with various types of exception/ptrs.

Refs https://github.com/scylladb/scylla/pull/10868

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

### CODE REVIEW

File: include/seastar/core/abort_source.hh [hunk 1]
```diff
@@ -136,20 +147,17 @@ public:
...
-            subscription& s = subs->front();
-            s.unlink();
-            s.on_abort(ex);
-        }
+    void request_abort() noexcept {
+        do_request_abort(std::nullopt);
+    }
+
+    void request_abort_ex(std::exception_ptr ex) noexcept {
+        do_request_abort(std::make_optional(std::move(ex)));
```
Comment (avikivity) [inline include/seastar/core/abort_source.hh:155]: Ugh, that is sad.

Comment (avikivity) [reply include/seastar/core/abort_source.hh:155]: Please add doxygen comments to the new function.

**Comment by bhalevy** [comment]
In v2 (f84b9a15): added doxygen comments for the new functions.
- rebased

File: include/seastar/core/abort_source.hh (hunk 1 above)
Comment (bhalevy) [reply include/seastar/core/abort_source.hh:160]: done in v2

**Pull request merged**